src/ScheduledTasks/Entity/Schedule.cs

```diff
@@ -0,0 +1,348 @@
+        TimeSpan interval = scheduleConfig.Interval;
+
+        if (executionToken != this.State.ExecutionToken)
+        {
+            this.logger.ScheduleRunCancelled(scheduleConfig.ScheduleId, executionToken);
+            return;
+        }
+
+        if (this.State.Status != ScheduleStatus.Active)
+        {
+            string errorMessage = "Schedule must be in Active status to run.";
+            Exception exception = new InvalidOperationException(errorMessage);
+            this.logger.ScheduleOperationError(scheduleConfig.ScheduleId, nameof(this.RunSchedule), errorMessage);
+            throw exception;
+        }
+
+        // if endat is set and time now is past endat, do not run
+        if (scheduleConfig.EndAt.HasValue && DateTimeOffset.UtcNow > scheduleConfig.EndAt.Value)
+        {
+            this.logger.ScheduleRunCancelled(scheduleConfig.ScheduleId, executionToken);
+            this.State.NextRunAt = null;
+
+            context.SignalEntity(
+                new EntityInstanceId(nameof(Schedule), scheduleConfig.ScheduleId),
+                "delete",
+                this.State.ExecutionToken);
+
+            return;
+        }
+
+        this.State.NextRunAt = this.DetermineNextRunTime(scheduleConfig);
+
+        DateTimeOffset currentTime = DateTimeOffset.UtcNow;
+
+        if (this.State.NextRunAt!.Value <= currentTime)
+        {
+            this.StartOrchestration(context, this.State.NextRunAt!.Value);
+            this.State.LastRunAt = this.State.NextRunAt!.Value;
+            this.State.NextRunAt = null;
+            this.State.NextRunAt = this.DetermineNextRunTime(scheduleConfig);
+        }
+
+        context.SignalEntity(
+            new EntityInstanceId(
+                nameof(Schedule),
+                this.State.ScheduleConfiguration.ScheduleId),
+            nameof(this.RunSchedule),
+            this.State.ExecutionToken,
+            new SignalEntityOptions { SignalTime = this.State.NextRunAt.Value });
+    }
+
+    void StartOrchestration(TaskEntityContext context, DateTimeOffset scheduledRunTime)
+    {
+        try
+        {
+            string? instanceId = this.State.ScheduleConfiguration?.OrchestrationInstanceId;
+            StartOrchestrationOptions startOrchestrationOptions;
+
+            if (string.IsNullOrEmpty(instanceId))
+            {
+                // Generate unique instance ID based on schedule name and current time
+                instanceId = $"{this.State.ScheduleConfiguration!.ScheduleId}-{scheduledRunTime:o}";
+                startOrchestrationOptions = new StartOrchestrationOptions(instanceId);
+            }
+            else
+            {
+                // Use configured instance ID which will prevent concurrent runs
+                startOrchestrationOptions = new StartOrchestrationOptions(instanceId);
+            }
+
+            this.logger.ScheduleOperationInfo(
+                this.State.ScheduleConfiguration!.ScheduleId,
```

Comment:
Isn't there a risk of a NRE here? Line 273 suggests that a null `ScheduleConfiguration` is possible, and if it's null there, it will be null here too.